File: frontend/src/utils/registrationValidation.js
const STUDENT_NUMBER_RE = /^\d{8}$/
const NWU_EMAIL_RE = /^(\d{8})@mynwu\.ac\.za$/i
const PASSWORD_RE = /^(?=.*[A-Za-z])(?=.*\d).{8,}$/

export const NWU_EMAIL_DOMAIN = '@mynwu.ac.za'

export function expectedNwuEmail(studentNumber) {
  return `${studentNumber.trim()}${NWU_EMAIL_DOMAIN}`
}

export function validateStudentNumber(studentNumber) {
  const value = studentNumber.trim()
  if (!value) return 'Enter your 8-digit NWU student number (e.g. 48277444).'
  if (!STUDENT_NUMBER_RE.test(value)) {
    return 'Student number must be exactly 8 digits (e.g. 48277444).'
  }
  return null
}

export function validateNwuEmail(email, studentNumber) {
  const normalized = email.trim().toLowerCase()
  const student = studentNumber.trim()
  if (!normalized) return 'Enter your NWU email (e.g. 48277444@mynwu.ac.za).'
  const match = NWU_EMAIL_RE.exec(normalized)
  if (!match) return `Email must be your student number followed by ${NWU_EMAIL_DOMAIN}`
  if (match[1] !== student) {
    return `Email must match your student number (e.g. ${expectedNwuEmail(student)})`
  }
  return null
}

export function validatePassword(password) {
  if (!password) return 'Password is required.'
  if (password.length < 8) return 'Password must be at least 8 characters.'
  if (!PASSWORD_RE.test(password)) {
    return 'Password must include at least one letter and one number.'
  }
  return null
}

export function validateUsername(username) {
  let value = username.trim().toLowerCase()
  if (value.includes(' ')) {
    value = value.split(/\s+/).join('.')
  }
  if (!value) return { error: 'Username is required.', username: value }
  if (!value.includes('.')) {
    return {
      error: 'Use a username like firstname.lastname (e.g. karabo.mabena).',
      username: value,
    }
  }
  return { error: null, username: value }
}

File: backend/utils/registration_validation.py
import re

STUDENT_NUMBER_RE = re.compile(r"^\d{8}$")
NWU_EMAIL_RE = re.compile(r"^(\d{8})@mynwu\.ac\.za$", re.IGNORECASE)
PASSWORD_RE = re.compile(r"^(?=.*[A-Za-z])(?=.*\d).{8,}$")

NWU_EMAIL_DOMAIN = "@mynwu.ac.za"


def normalize_student_number(value):
    return (value or "").strip()


def normalize_nwu_email(value):
    return (value or "").strip().lower()


def expected_nwu_email(student_number):
    return f"{student_number}{NWU_EMAIL_DOMAIN}"


def validate_student_number(student_number):
    student_number = normalize_student_number(student_number)
    if not student_number:
        return "STUDENT_NUMBER_REQUIRED"
    if not STUDENT_NUMBER_RE.match(student_number):
        return "STUDENT_NUMBER_INVALID"
    return None


def validate_nwu_email(email, student_number):
    email = normalize_nwu_email(email)
    student_number = normalize_student_number(student_number)
    if not email:
        return "EMAIL_REQUIRED"
    match = NWU_EMAIL_RE.match(email)
    if not match:
        return "EMAIL_INVALID"
    if match.group(1) != student_number:
        return "EMAIL_STUDENT_MISMATCH"
    return None


def validate_password(password):
    if not password:
        return "PASSWORD_REQUIRED"
    if len(password) < 8:
        return "PASSWORD_TOO_SHORT"
    if not PASSWORD_RE.match(password):
        return "PASSWORD_WEAK"
    return None


def validate_registration(username, email, password, student_number):
    student_number = normalize_student_number(student_number)
    email = normalize_nwu_email(email)

    err = validate_student_number(student_number)
    if err:
        return err
    err = validate_nwu_email(email, student_number)
    if err:
        return err
    err = validate_password(password)
    if err:
        return err
    if not (username or "").strip():
        return "USERNAME_REQUIRED"
    if "." not in username.strip():
        return "USERNAME_INVALID"
    return None
